proofs-plugin/src/main/kotlin/arrow/meta/plugins/proofs/phases/quotes/GenerateGivenSupportingFunctions.kt

```diff
@@ -3,88 +3,136 @@ package arrow.meta.plugins.proofs.phases.quotes
+              $name(${valueParameters.renderAsArguments()}) 
+            """.trimIndent()
+        }
+        is ClassDescriptor -> {
+          if (unsubstitutedPrimaryConstructor?.valueParameters?.isNotEmpty() == true) {
+            """ 
+            package ${f.containingPackage()}
+            @arrow.CompileTime
+            internal fun ${declaredTypeParameters.render()} $name(${unsubstitutedPrimaryConstructor?.valueParameters?.renderParameters()}): $defaultType = 
+              $name(${unsubstitutedPrimaryConstructor?.valueParameters?.renderAsArguments()}) 
+            """.trimIndent()
+          } else ""
+        }
+        else -> ""
       }
+    s
+  }
 
-      val `(paramsWithGiven)` = `(params)` + `(givenParams)`
-      """
-      public fun $`(unconstrainedTypeParams)` $receiver$name $`(paramsWithGiven)`$returnType =
-          ${runScope(this, `(givenParams)`)}
-      """.function(null).value
-    }
-  }, separator = "\n")
+private fun List<TypeParameterDescriptor>.render(): String =
+  if (isEmpty()) ""
+  else joinToString(prefix = "<", postfix = ">") { it.name.asString() }
 
-private fun ElementScope.runScope(namedFunction: NamedFunction, scopedList: ScopedList<KtParameter>): Scope<KtExpression> {
-  val body = namedFunction.body
-  return if (body != null)
-    scopedList.value.fold(body.toString().expression) { acc, parameter ->
-      """
-      with<${parameter.typeReference?.text}, ${namedFunction.returnType.copy(prefix = "")}>(${parameter.name}) { $acc }
-      """.expression
-    }
-  else Scope.empty<KtExpression>()
+private fun ReceiverParameterDescriptor?.render(): String {
+  return this?.value?.type?.toString()?.let { "$it."} ?: ""
 }
 
-private fun KtTypeParameter.givenConstrain(): TypeReference? =
-  if (containsGivenConstrain()) TypeReference(extendsBound) else null
+private fun List<ValueParameterDescriptor>.renderParameters(): String =
+  joinToString {
+    val context = it.contextualAnnotations().firstOrNull()
+    if (it.isProof() && context != null) "@$context ${it.name}: ${it.type} = TODO(\"Compile time replaced\")"
+    else "${it.name}: ${it.type}"
+  } + ", unit: Unit = Unit"
 
-private fun KtNamedFunction.containsGivenConstrain(): Boolean =
-  typeParameters.any { typeParameter ->
-    typeParameter.containsGivenConstrain()
-  }
+private fun List<ValueParameterDescriptor>.renderAsArguments(): String =
+  joinToString { it.name.asString() }
 
-private fun KtTypeParameter.containsGivenConstrain(): Boolean =
-  extendsBound?.annotationEntries?.any { it.text.matches(givenAnnotation) } ?: false
\ No newline at end of file
+private tailrec fun ModuleDescriptor.declarationsWithGivenArguments(
+  acc: List<DeclarationDescriptor> = emptyList(),
+  packages: List<FqName> = listOf(FqName.ROOT),
+  skipPacks: Set<FqName> = setOf(
```

Comment:
This is an optmisation technique? Since these packages can never contain any of our annotations?

Reply:
Yes, there is currently no automatic way in the compiler to scan the classpath for annotated descriptors, and this ensures we don't recurse down paths where we know there are no proofs or we explicitly want to ignore any proofs people are trying to project over core packages